week2/homework/src/index.js

```diff
@@ -1,3 +1,183 @@
+    .then(JSON.parse)
+    .catch(() => []);
+}
+
+function storeTodos(todos) {
+  return writeFilePromisified(TODOS_PATH, JSON.stringify(todos, null, 2));
+}
+
+function resetTodos() {
+  return unlinkPromisified(TODOS_PATH).catch(() => {
+    console.log('no todos to reset, you might want to **add** one first');
+    console.log(helpMessages.help);
+  });
+}
+
+const helpMessages = {
+  add: 'use: node . add "something to do"',
+  remove: 'use: node . remove <index>',
+  list: 'use: node . list',
+  done: 'use: node . done <index>',
+  update: 'use: node . update <index> "something to do instead"',
+  reset: 'use: node . reset',
+
+  help: boxt(
+    `${chalk.cyan('>>')} Available commands:
+node . add 'something to do'
+${chalk.inverse.green('Adds a todo to your todo list.')}
+node . remove <index>
+${chalk.inverse.green('Removes a todo, specified by the index.')}
+node . list
+${chalk.inverse.green('Lists all todos.')}
+node . done <index>
+${chalk.inverse.green('Marks a todo as done, specified by the index.')}
+node . update <index> 'new todo'
+${chalk.inverse.green(
+      'Replaces a todo specified by the index withe a new todo.'
+    )}
+node . reset
+${chalk.inverse.green('Resets the todos file.')}
+node . OR node . help
+${chalk.inverse.green('Shows the available commands (this).')}`,
+    {
+      theme: 'round',
+      color: 'green',
+      align: 'left'
+    }
+  )
+};
+
+async function main() {
+  const userIndexInput = parseInt(args[1]);
+  const allTodos = await readJSON();
+
+  function addTodo() {
+    const userTodoInput = args[1];
+    if (userTodoInput === undefined) {
+      console.log(
+        chalk.red('Error while adding a todo, see this:\n'),
+        helpMessages.add
+      );
+    } else {
+      allTodos.push({
+        task: userTodoInput,
+        done: 'not done'
+      });
+      storeTodos(allTodos);
+      console.log(`added todo: '${chalk.cyan(args[1])}'`);
+    }
+  }
+
+  function removeTodo() {
+    if (userIndexInput >= allTodos.length) {
+      console.log(`${chalk.red("didn't find a todo with that index")}`);
+      console.log(helpMessages.remove);
+    }
+    allTodos.forEach((todo, index) => {
+      if (userIndexInput === index) {
+        allTodos.pop(index);
+        storeTodos(allTodos);
+        console.log(`successfully removed '${chalk.cyan(todo.task)}'`);
+      }
+    });
+  }
+
+  function listTodos() {
+    if (allTodos.length === 0) {
+      console.log('no todos yet, start by adding one');
+      console.log(helpMessages.add);
+    } else {
+      console.log('your todos are:');
+      allTodos.forEach((todo, index) => {
+        console.log(
+          `  ${chalk.cyan(index)}- ${todo.task} ${
+            todo.done === 'done'
+              ? chalk.green.bold(todo.done)
+              : chalk.magenta.bold(todo.done)
+          }`
+        );
+      });
+    }
+  }
+
+  function markTodoAsDone() {
+    allTodos.forEach((todo, index) => {
```

Comment:
Why do you iterate over the whole array to update 1 item? The whole point of an array is that you can access item by index. Also, the assignment is to remove and update items using 1-based index and not zero-based.

Reply:
All fixed too :heavy_check_mark: .